Features/receipt form

viennguyen24 opened the pull request:
**Issue #39:** Form for image scan and schema validation for user register / login page

**Branch name:** features/receipt-form

**PR #39:** Implemented UI for receipt-form, display mock receipt data and allows user to manually change, update, and delete data. Changed hand-written form validation into schema validation with Zod for user register / login

**Libraries:** React-hook-form, zod, lucide icon react


**Mock data**: Please refer to #47  for mock data

**Demo receipt form**
[https://drive.google.com/file/d/1EExC536tGbBr4BF4lFR29h0NkUIm2Xw6/view?usp=drive_link](url)




**Files involved: client/src/pages/ReceiptForm.tsx**

**Overview**
This is a form page where users can receive data about scanned ingredients and manually update the ingredients / add new ones. 
1. Backend sends data about ingredients that users scanned from their receipt. Each item in the data includes:
- Name of the ingredient
- Price 
- Quantity 
- Suggested ingredients (in case the ingredient was mismatched or irrelevant)

2. User can manually adjust the name / price / quantity in each input field or delete the ingredient by clicking the **Trash icon** button. If the ingredient was mismatched or irrelevant (for example the ingredient was chicken breast but the name was only chicken), then user can click on the **Pencil icon** button where they will receive a list of ingredients that might be the right one.

3. Once all ingredients are selected, users can click on submit changes and send the ingredients back to the backend. The data should look like this: 
```typescript
{receipts: 
     [{name: 'pumpkin vanilla', quantity: '1', price: '165.00'},
      {name: 'tomato & cheese', quantity: '1', price: '185.00' },]
}
```

**Work to be done**
- [ ] Calling backend to retrieve receipts data 
- [x] Creating a state to manage current receipts and load current receipts, as well as to store data retrieved
- [ ]  Break down some code into separate component to reuse in the future (if needed) 
- [ ] Bring hard-fixed data / types outside for reusability
- [ ] Update UI and implement responsiveness / accessibility

**Schema validation for authentication page**
```typescript

\\\ Register.tsx
const UserRegister = z.object({
  username: z.string().min(4, { message: "Username must be at least 4 characters" }),
  email: z.string().email({ message: "Please enter a valid email"}),
  password: z.string().min(8, { message: "Password must be at least 8 characters" }),
  reEnterPassword: z.string().min(8, { message: "Password must be at least 8 characters" }),
}).refine(data => data.password === data.reEnterPassword, {
  message: "You must re-enter your password correctly",
  path: ["reEnterPassword"], 
});

const RequestRegister = z.object({
  name: z.string(),
  email: z.string(),
  password: z.string(),
});

type UserRegisterType = z.infer<typeof UserRegister>;
type RequestRegisterType = z.infer<typeof RequestRegister>;
```

```typescript
\\Login.tsx

const UserLogin = z
  .object({
    email: z.string().email({ message: 'Please enter a valid email' }),
    password: z.string().min(8, { message: 'Please enter a password with 8 characters minimum' }),
  })
  .strict().required({ email: true, password: true });
type UserLoginType = z.infer<typeof UserLogin>;
```

viennguyen24 committed:
schema validation added for login page

viennguyen24 committed:
new receipt page with field array added

viennguyen24 committed:
fixed form submit type errors, added state for receipts and update ui to show state changes

viennguyen24 committed:
UI completed for receipt form,need to modularize component for reusability

viennguyen24 committed:
fix bugs related to commits

viennguyen24 committed:
format code and fixed some styling

hungngodev reviewed (changes requested):
Change the login to Login file

hungngodev commented on client/src/pages/mockReceipt.ts:
```diff
@@ -0,0 +1,42 @@
+      quantity: '1',
+    },
+  ],
+};
```
Move this mock file into the ReceiptForm.tsx. Also the Receipt Form should be placed in the component folder not pages folder.

hungngodev commented on client/src/pages/login.tsx:
```diff
@@ -9,25 +9,29 @@ import {
+    email: z.string().email({ message: 'Please enter a valid email' }),
+    password: z.string().min(8, { message: 'Please enter a password with 8 characters minimum' }),
+  })
+  .strict();
```
I have not seen the same for the Register page. Could you upload to this same PR?

viennguyen24 replied:
Hi Hung, I uploaded both the Register and Login page at commit #[b2e3a42](https://github.com/hungngodev/Flavorie/pull/49/commits/b2e3a427abeaa7b2ec569b5c77b5d93dd927ff45)!

viennguyen24 committed:
populate the receipt state with mock data and fix type errors between field array and zod object

viennguyen24 committed:
removed unecessary files from pr

viennguyen24 committed:
remove remaining unecessary files

viennguyen24 committed:
Rename mockReceipt.ts to MockReceipt.ts

viennguyen24 committed:
remove redefinition of mock data

import mockReceipts data straight from MockReceipt data file instead of redefining it in the ReceiptForm file

hungngodev commented on client/src/pages/MockReceipt.ts:
Could you move this file with the ReceiptForm.tsx to the components folder?

viennguyen24 replied:
> Could you move this file with the ReceiptForm.tsx to the components folder?

Check commit #[7820fba](https://github.com/hungngodev/Flavorie/pull/49/commits/7820fba3cfa249047c060da06faae58a18eb0da9) !

viennguyen24 committed:
Merge branch 'main' into features/receipt-form

viennguyen24 committed:
schema validation added for both login and register page

viennguyen24 committed:
schema validation added for login and register page

viennguyen24 committed:
schema validation added for login and register page

viennguyen24 committed:
changed location of ReceiptForm file,brought mock data inside file

viennguyen24 commented on client/src/pages/ReceiptForm.tsx:
This file is moved into components/ingredients folder

hungngodev committed:
Delete client/src/pages/ReceiptForm.tsx

hoatuyet423 commented:
- Confused, is this branch suppose to resolve 2 issues? Only one is mentioned in the description. If yes, then should not do this in the future, only 1 PR per 1 branch per issue. 
- A lot of styling in the code is really inconsistent. For example how your css is structures is different in different files and places in your code. 
- I didnt read much about the html as it will mostly shown in the UI, but leave a few comments here and there
- Can you provide a video of interaction with the UI and what it does?
- NIT: I feel like the huge space in between the item card is making it feel empty. Can make the order summary card bigger to balance this out 
- Overall UI look nice and clean, but feel like the code is still pretty confused to follow .-. Good works tho :>

viennguyen24 committed:
added update toggle and select component to display suggestions for each item

viennguyen24 committed:
gMerge branch 'features/receipt-form' of https://github.com/hungngodev/Flavorie into features/receipt-form

viennguyen24 committed:
update form based on new back-end data sample

viennguyen24 committed:
removed unecessary state, fixed bugs about adding / deleting and submitting form

viennguyen24 committed:
ui updated, replaced unecessary containers with chakra card component, fixed bugs relating to form, added image field

viennguyen24 committed:
change receipt form and receipt field into seperate component, issues on passing fields and watch component

viennguyen24 committed:
added watch prop to fix bug about reading field value

viennguyen24 committed:
resolve form auto submitting from menubutton, fixed type errors from passing field props, integrated custom input components

hoatuyet423 commented:
- resolved conflict 
- demo link doesnt work
